### PR TITLE
tls: add client-side CA authenticator

### PR DESCRIPTION
**Pull request by samoht**

### CODE REVIEW

**Comment by hannesm** [comment]
this is good imho, though two nits:
- it should be documented (I'm not sure in which scenarios conduit is used, but previously no certificate was validated -- now it is validated against the OS trust anchors -- this works for the public internet, but won't for private network setups with your custom CA)
- I'd appreciate a corresponding change to conduit-mirage (I briefly looked into the details, and it seems that getting a Mirage_clock.PCLOCK (for ca-certs-nss / Ca_certs_nss.Make) from within conduit_mirage is rather painful (will need API / functor changes, which need to be reflected in the mirage utility)

TL;DR: best document that now OS trust anchors are used (and wait if someone has a usecase where they need their own (or no) trust anchors) :) and revising conduit_mirage would be appreciated.

**Comment by samoht** [comment]
Updated with changes to `conduit-mirage` too. It seems that only `mirage-skeleton` is using client-side conduit, so I can fix that after the release.

**Comment by samoht** [comment]
Conduit_mirage is really terrible :p maybe it's a good time to simplify this again (but in a different PR).

File: src/conduit-lwt-unix/conduit_lwt_unix.mli [hunk 1]
```diff
@@ -166,7 +166,8 @@ val init :
 (** [init ?src ?tls_own_key ()] will initialize a Unix conduit that binds to the
     [src] interface if specified. If TLS server connections are used, then
     [tls_server_key] must contain a valid certificate to be used to advertise a
-    TLS connection *)
+    TLS connection. The certificate is validated against the OS trust anchors
```
Comment (avsm) [inline src/conduit-lwt-unix/conduit_lwt_unix.mli:169]: Can't we pass through the `authenticator` function here as an optional arg to init? That would allow it to be overridden for custom CAs, but the default would work for public anchors.

**Comment by samoht** [comment]
@avsm I've pushed a commit to select the correct authenticator.

I am not totally sure how to do this with `ocaml-ssl` so it's TLS-only.

File: src/conduit-lwt-unix/conduit_lwt_unix.mli [hunk 2]
```diff
@@ -160,13 +160,16 @@ val default_ctx : ctx
...
   unit ->
   ctx io
 (** [init ?src ?tls_own_key ()] will initialize a Unix conduit that binds to the
     [src] interface if specified. If TLS server connections are used, then
     [tls_server_key] must contain a valid certificate to be used to advertise a
-    TLS connection *)
+    TLS connection.
+
+    The certificate is validated using [tls_authenticator]. By default, the
+    validation is using the OS trust anchors. *)
```
Comment (hannesm) [inline src/conduit-lwt-unix/conduit_lwt_unix.mli:172]:
```suggestion
    validation is using the {{:https://github.com/mirage/cacerts}OS trust anchors}. *)
```

File: src/conduit-mirage/conduit_mirage.mli [hunk 3]
```diff
@@ -93,8 +93,15 @@ type server = [ tcp_server | vchan_server | server tls_server ]
...
 
-val client : Conduit.endp -> client Lwt.t
-(** Resolve a conduit endpoint into a client configuration. *)
+module Client (P : Mirage_clock.PCLOCK) : sig
+  val resolve :
+    ?tls_authenticator:X509.Authenticator.t -> Conduit.endp -> client Lwt.t
+  (** Resolve a conduit endpoint into a client configuration.
+
+      The certificate is validated using [tls_authenticator]. By default, the
+      validation is using the trust anchors extracted from Mozilla's NSS
```
Comment (hannesm) [inline src/conduit-mirage/conduit_mirage.mli:102]:
```suggestion
      validation is using the {{:https://github.com/mirage/ca-certs-nss}trust anchors extracted from Mozilla's NSS}.
```

File: src/conduit-mirage/conduit_mirage.mli [hunk 4]
```diff
@@ -93,8 +93,15 @@ type server = [ tcp_server | vchan_server | server tls_server ]
...
+      certdata.txt package. *)
```
Comment (hannesm) [inline src/conduit-mirage/conduit_mirage.mli:103]:
```suggestion
```

**Comment by hannesm** [comment]
Thanks @samoht -- I think this is good (moving conduit-lwt and conduit-mirage on the same level).

File: src/conduit-lwt-unix/conduit_lwt_unix.mli [hunk 5]
```diff
@@ -160,13 +160,17 @@ val default_ctx : ctx
...
   unit ->
   ctx io
 (** [init ?src ?tls_own_key ()] will initialize a Unix conduit that binds to the
     [src] interface if specified. If TLS server connections are used, then
     [tls_server_key] must contain a valid certificate to be used to advertise a
-    TLS connection *)
+    TLS connection.
+
+    The certificate is validated using [tls_authenticator]. By default, the
+    validation is using the {{:https://github.com/mirage/cacerts} OS trust
```
Comment (avsm) [inline src/conduit-lwt-unix/conduit_lwt_unix.mli:172]:
```suggestion
    validation is using the {{:https://github.com/mirage/ca-certs} OS trust
```

**Review by avsm — APPROVED** [review]
lgtm

**Comment by avsm** [comment]
(needs a changes entry)

**Comment by samoht** [comment]
I've just added a CHANGE entry
